tket/src/Circuit/basic_circ_manip.cpp

```diff
@@ -80,18 +80,47 @@ Vertex Circuit::add_op<unsigned>(
     const Op_ptr& gate, const std::vector<unsigned>& args,
     std::optional<std::string> opgroup) {
   op_signature_t sig = gate->get_signature();
-  if (sig.size() != args.size()) {
+
+  // check if there is wasm in the signature
+  bool sig_contains_wasm = false;
+  for (EdgeType e : sig) {
+    if (e == EdgeType::WASM) {
+      sig_contains_wasm = true;
+      break;
+    }
+  }
+
+  // sig without wasm
+  if (sig.size() != args.size() && !sig_contains_wasm) {
     throw CircuitInvalidity(
         std::to_string(args.size()) + " args provided, but " +
         gate->get_name() + " requires " + std::to_string(sig.size()));
   }
+
+  // sig with wasm
```

Comment:
Feel like this code has appeared somewhere else -- can we refactor to avoid duplication?